scorch crashes on writeDicts

Scorch crashes if you have a mapping where the same term is included in both a field where we don't include the term vectors and another field where we do.

When that happens the frequency of the term is 2 but the locs array only contains one item inside the writeDicts function when processing  the term `something`.

Here is the example code and the output when running it (in macOS High Sierra):

```
package main

import (
	"log"

	"github.com/blevesearch/bleve"
	"github.com/blevesearch/bleve/analysis/analyzer/keyword"
	"github.com/blevesearch/bleve/analysis/analyzer/standard"
	"github.com/blevesearch/bleve/index/scorch"
)

func main() {
	mp := bleve.NewIndexMapping()
	keywordMapping := bleve.NewTextFieldMapping()
	keywordMapping.Analyzer = keyword.Name
	keywordMapping.Store = false
	keywordMapping.IncludeTermVectors = false
	textMapping := bleve.NewTextFieldMapping()
	textMapping.Analyzer = standard.Name
	textMapping.Store = false
	docMapping := bleve.NewDocumentStaticMapping()
	docMapping.AddFieldMappingsAt("keyword", keywordMapping)
	docMapping.AddFieldMappingsAt("text", textMapping)
	mp.AddDocumentMapping("doc", docMapping)
	mp.DefaultType = "doc"
	mp.DefaultMapping = docMapping
	mp.DefaultAnalyzer = standard.Name
	mp.DocValuesDynamic = false
	mp.StoreDynamic = false
	mp.IndexDynamic = false

	idx, err := bleve.NewUsing("idx_data", mp, scorch.Name, scorch.Name, nil)
	if err != nil {
		log.Fatalln(err)
	}

	data := map[string]string{
		"keyword": "something",
		"text":    "A sentence that includes something within.",
	}

	idx.Index("id", data)
}
```

```
panic: runtime error: index out of range

goroutine 1 [running]:
test/crashtest/vendor/github.com/blevesearch/bleve/index/scorch/segment/zap.(*interim).writeDicts(0xc4200c4a00, 0x6, 0x0, 0x0, 0x4, 0x4, 0xc420077b48)
	/Users/me/go/src/test/crashtest/vendor/github.com/blevesearch/bleve/index/scorch/segment/zap/new.go:643 +0x1b48
test/crashtest/vendor/github.com/blevesearch/bleve/index/scorch/segment/zap.(*interim).convert(0xc4200c4a00, 0xc42010b300, 0xc4200c4a00, 0xc, 0xc4201aea17, 0x3, 0x8, 0x0, 0xc420082000)
	/Users/me/go/src/test/crashtest/vendor/github.com/blevesearch/bleve/index/scorch/segment/zap/new.go:247 +0x386
test/crashtest/vendor/github.com/blevesearch/bleve/index/scorch/segment/zap.AnalysisResultsToSegmentBase(0xc42000e390, 0x1, 0x1, 0x400, 0x1, 0x1, 0x0)
	/Users/me/go/src/test/crashtest/vendor/github.com/blevesearch/bleve/index/scorch/segment/zap/new.go:53 +0x127
test/crashtest/vendor/github.com/blevesearch/bleve/index/scorch.(*Scorch).Batch(0xc4200aaa80, 0xc4201ec060, 0x0, 0x0)
	/Users/me/go/src/test/crastest/vendor/github.com/blevesearch/bleve/index/scorch/scorch.go:312 +0x7b2
test/crashtest/vendor/github.com/blevesearch/bleve/index/scorch.(*Scorch).Update(0xc4200aaa80, 0xc4200328c0, 0x13bf480, 0xc4201ae9c0)
	/Users/me/go/src/test/crashtest/vendor/github.com/blevesearch/bleve/index/scorch/scorch.go:246 +0xd9
test/crashtest/vendor/github.com/blevesearch/bleve.(*indexImpl).Index(0xc42018e150, 0x1441d90, 0x2, 0x13bf480, 0xc4201ae9c0, 0x0, 0x0)
	/Users/me/go/src/test/crashtest/crashtest/github.com/blevesearch/bleve/index_impl.go:257 +0x228
main.main()
	/Users/me/go/src/test/crashtest/crash.go:42 +0x470
```

The workaround is to include the term vectors for all the fields.
This panic only happens when using scorch, I have tried both moss and boltdb and it works without a problem.

--- a/index/scorch/scorch_test.go
+++ b/index/scorch/scorch_test.go
@@ -25,9 +25,12 @@ import (
 	"time"
 
 	"github.com/blevesearch/bleve/analysis"
+	"github.com/blevesearch/bleve/analysis/analyzer/keyword"
+	"github.com/blevesearch/bleve/analysis/analyzer/standard"
 	regexpTokenizer "github.com/blevesearch/bleve/analysis/tokenizer/regexp"
 	"github.com/blevesearch/bleve/document"
 	"github.com/blevesearch/bleve/index"
+	"github.com/blevesearch/bleve/mapping"
 )
 
 func DestroyTest() error {
@@ -1706,4 +1709,56 @@ func TestIndexDocumentVisitFieldTermsWithMultipleFieldOptions(t *testing.T) {
 		t.Fatal(err)
 	}
 
+}
+
+func TestAllFieldWithDifferentTermVectorsEnabled(t *testing.T) {
+	// Based on https://github.com/blevesearch/bleve/issues/895 from xeizmendi
+	mp := mapping.NewIndexMapping()
+
+	keywordMapping := mapping.NewTextFieldMapping()
+	keywordMapping.Analyzer = keyword.Name
+	keywordMapping.IncludeTermVectors = false
+	keywordMapping.IncludeInAll = true
+
+	textMapping := mapping.NewTextFieldMapping()
+	textMapping.Analyzer = standard.Name
+	textMapping.IncludeTermVectors = true
+	textMapping.IncludeInAll = true
+
+	docMapping := mapping.NewDocumentStaticMapping()
+	docMapping.AddFieldMappingsAt("keyword", keywordMapping)
+	docMapping.AddFieldMappingsAt("text", textMapping)
+
+	mp.DefaultMapping = docMapping
+
+	_ = os.RemoveAll(testConfig["path"].(string))
+	analysisQueue := index.NewAnalysisQueue(1)
+	idx, err := NewScorch("storeName", testConfig, analysisQueue)
+	if err != nil {
+		log.Fatalln(err)
+	}
+	err = idx.Open()
+	if err != nil {
+		t.Errorf("error opening index: %v", err)
+	}
+	defer func() {
+		_ = idx.Close()
+		_ = os.RemoveAll(testConfig["path"].(string))
+	}()
+
+	data := map[string]string{
+		"keyword": "something",
+		"text":    "A sentence that includes something within.",
+	}
+
+	doc := document.NewDocument("1")
+	err = mp.MapDocument(doc, data)
+	if err != nil {
+		t.Errorf("error mapping doc: %v", err)
+	}
+
+	err = idx.Update(doc)
+	if err != nil {
+		t.Errorf("Error updating index: %v", err)
+	}
 }

--- a/index/scorch/segment/zap/build.go
+++ b/index/scorch/segment/zap/build.go
@@ -20,7 +20,7 @@ import (
 	"os"
 )
 
-const Version uint32 = 9
+const Version uint32 = 10
 
 const Type string = "zap"
 

--- a/index/scorch/segment/zap/merge.go
+++ b/index/scorch/segment/zap/merge.go
@@ -427,17 +427,30 @@ func mergeTermFreqNormLocs(fieldsMap map[string]uint16, term []byte, postItr *Po
 		}
 
 		if len(locs) > 0 {
+			numBytesLocs := 0
 			for _, loc := range locs {
-				if cap(bufLoc) < 5+len(loc.ArrayPositions()) {
-					bufLoc = make([]uint64, 0, 5+len(loc.ArrayPositions()))
+				ap := loc.ArrayPositions()
+				numBytesLocs += totalUvarintBytes(uint64(fieldsMap[loc.Field()]-1),
+					loc.Pos(), loc.Start(), loc.End(), uint64(len(ap)), ap)
+			}
+
+			err = locEncoder.Add(hitNewDocNum, uint64(numBytesLocs))
+			if err != nil {
+				return 0, 0, 0, nil, err
+			}
+
+			for _, loc := range locs {
+				ap := loc.ArrayPositions()
+				if cap(bufLoc) < 5+len(ap) {
+					bufLoc = make([]uint64, 0, 5+len(ap))
 				}
 				args := bufLoc[0:5]
 				args[0] = uint64(fieldsMap[loc.Field()] - 1)
 				args[1] = loc.Pos()
 				args[2] = loc.Start()
 				args[3] = loc.End()
-				args[4] = uint64(len(loc.ArrayPositions()))
-				args = append(args, loc.ArrayPositions()...)
+				args[4] = uint64(len(ap))
+				args = append(args, ap...)
 				err = locEncoder.Add(hitNewDocNum, args...)
 				if err != nil {
 					return 0, 0, 0, nil, err

--- a/index/scorch/segment/zap/new.go
+++ b/index/scorch/segment/zap/new.go
@@ -190,7 +190,7 @@ type interimStoredField struct {
 type interimFreqNorm struct {
 	freq    uint64
 	norm    float32
-	hasLocs bool
+	numLocs int
 }
 
 type interimLoc struct {
@@ -446,7 +446,7 @@ func (s *interim) processDocument(docNum uint64,
 				interimFreqNorm{
 					freq:    uint64(tf.Frequency()),
 					norm:    norm,
-					hasLocs: len(tf.Locations) > 0,
+					numLocs: len(tf.Locations),
 				})
 
 			if len(tf.Locations) > 0 {
@@ -632,18 +632,28 @@ func (s *interim) writeDicts() (fdvIndexOffset uint64, dictOffsets []uint64, err
 				freqNorm := freqNorms[freqNormOffset]
 
 				err = tfEncoder.Add(docNum,
-					encodeFreqHasLocs(freqNorm.freq, freqNorm.hasLocs),
+					encodeFreqHasLocs(freqNorm.freq, freqNorm.numLocs > 0),
 					uint64(math.Float32bits(freqNorm.norm)))
 				if err != nil {
 					return 0, nil, err
 				}
 
-				for i := uint64(0); i < freqNorm.freq; i++ {
-					if len(locs) > 0 {
-						loc := locs[locOffset]
+				if freqNorm.numLocs > 0 {
+					numBytesLocs := 0
+					for _, loc := range locs[locOffset : locOffset+freqNorm.numLocs] {
+						numBytesLocs += totalUvarintBytes(
+							uint64(loc.fieldID), loc.pos, loc.start, loc.end,
+							uint64(len(loc.arrayposs)), loc.arrayposs)
+					}
 
-						err = locEncoder.Add(docNum, uint64(loc.fieldID),
-							loc.pos, loc.start, loc.end,
+					err = locEncoder.Add(docNum, uint64(numBytesLocs))
+					if err != nil {
+						return 0, nil, err
+					}
+
+					for _, loc := range locs[locOffset : locOffset+freqNorm.numLocs] {
+						err = locEncoder.Add(docNum,
+							uint64(loc.fieldID), loc.pos, loc.start, loc.end,
 							uint64(len(loc.arrayposs)))
 						if err != nil {
 							return 0, nil, err
@@ -655,7 +665,7 @@ func (s *interim) writeDicts() (fdvIndexOffset uint64, dictOffsets []uint64, err
 						}
 					}
 
-					locOffset++
+					locOffset += freqNorm.numLocs
 				}
 
 				freqNormOffset++
@@ -774,4 +784,27 @@ func encodeFieldType(f document.Field) byte {
 		fieldType = 'c'
 	}
 	return fieldType
+}
+
+// returns the total # of bytes needed to encode the given uint64's
+// into binary.PutUVarint() encoding
+func totalUvarintBytes(a, b, c, d, e uint64, more []uint64) (n int) {
+	n = numUvarintBytes(a)
+	n += numUvarintBytes(b)
+	n += numUvarintBytes(c)
+	n += numUvarintBytes(d)
+	n += numUvarintBytes(e)
+	for _, v := range more {
+		n += numUvarintBytes(v)
+	}
+	return n
+}
+
+// returns # of bytes needed to encode x in binary.PutUvarint() encoding
+func numUvarintBytes(x uint64) (n int) {
+	for x >= 0x80 {
+		x >>= 7
+		n++
+	}
+	return n + 1
 }

--- a/index/scorch/segment/zap/posting.go
+++ b/index/scorch/segment/zap/posting.go
@@ -18,6 +18,7 @@ import (
 	"bytes"
 	"encoding/binary"
 	"fmt"
+	"io"
 	"math"
 	"reflect"
 
@@ -539,7 +540,10 @@ func (i *PostingsIterator) nextAtOrAfter(atOrAfter uint64) (segment.Posting, err
 	rv.norm = math.Float32frombits(uint32(normBits))
 
 	if i.includeLocs && hasLocs {
-		// read off 'freq' locations, into reused slices
+		// prepare locations into reused slices, where we assume
+		// rv.freq >= "number of locs", since in a composite field,
+		// some component fields might have their IncludeTermVector
+		// flags disabled while other component fields are enabled
 		if cap(i.nextLocs) >= int(rv.freq) {
 			i.nextLocs = i.nextLocs[0:rv.freq]
 		} else {
@@ -548,13 +552,22 @@ func (i *PostingsIterator) nextAtOrAfter(atOrAfter uint64) (segment.Posting, err
 		if cap(i.nextSegmentLocs) < int(rv.freq) {
 			i.nextSegmentLocs = make([]segment.Location, rv.freq, rv.freq*2)
 		}
-		rv.locs = i.nextSegmentLocs[0:rv.freq]
-		for j := 0; j < int(rv.freq); j++ {
+		rv.locs = i.nextSegmentLocs[:0]
+
+		numLocsBytes, err := binary.ReadUvarint(i.locReader)
+		if err != nil {
+			return nil, fmt.Errorf("error reading location numLocsBytes: %v", err)
+		}
+
+		j := 0
+		startBytesRemaining := i.locReader.Len() // # bytes remaining in the locReader
+		for startBytesRemaining-i.locReader.Len() < int(numLocsBytes) {
 			err := i.readLocation(&i.nextLocs[j])
 			if err != nil {
 				return nil, err
 			}
-			rv.locs[j] = &i.nextLocs[j]
+			rv.locs = append(rv.locs, &i.nextLocs[j])
+			j++
 		}
 	}
 
@@ -597,11 +610,16 @@ func (i *PostingsIterator) nextBytes() (
 	if hasLocs {
 		startLoc := len(i.currChunkLoc) - i.locReader.Len()
 
-		for j := uint64(0); j < freq; j++ {
-			err := i.readLocation(nil)
-			if err != nil {
-				return 0, 0, 0, nil, nil, err
-			}
+		numLocsBytes, err := binary.ReadUvarint(i.locReader)
+		if err != nil {
+			return 0, 0, 0, nil, nil,
+				fmt.Errorf("error reading location nextBytes numLocs: %v", err)
+		}
+
+		// skip over all the location bytes
+		_, err = i.locReader.Seek(int64(numLocsBytes), io.SeekCurrent)
+		if err != nil {
+			return 0, 0, 0, nil, nil, err
 		}
 
 		endLoc := len(i.currChunkLoc) - i.locReader.Len()
@@ -659,17 +677,21 @@ func (i *PostingsIterator) nextDocNumAtOrAfter(atOrAfter uint64) (uint64, bool, 
 			}
 
 			// read off freq/offsets even though we don't care about them
-			freq, _, hasLocs, err := i.readFreqNormHasLocs()
+			_, _, hasLocs, err := i.readFreqNormHasLocs()
 			if err != nil {
 				return 0, false, err
 			}
 
 			if i.includeLocs && hasLocs {
-				for j := 0; j < int(freq); j++ {
-					err := i.readLocation(nil)
-					if err != nil {
-						return 0, false, err
-					}
+				numLocsBytes, err := binary.ReadUvarint(i.locReader)
+				if err != nil {
+					return 0, false, fmt.Errorf("error reading location numLocsBytes: %v", err)
+				}
+
+				// skip over all the location bytes
+				_, err = i.locReader.Seek(int64(numLocsBytes), io.SeekCurrent)
+				if err != nil {
+					return 0, false, err
 				}
 			}
 		}

